pkg/scaffold/resource/resource.go

```diff
@@ -54,38 +54,58 @@ type Resource struct {
 
 // Validate checks the Resource values to make sure they are valid.
 func (r *Resource) Validate() error {
-	if len(r.Group) == 0 {
+	if r.isGroupEmpty() {
 		return fmt.Errorf("group cannot be empty")
 	}
-	if len(r.Version) == 0 {
+	if r.isVersionEmpty() {
 		return fmt.Errorf("version cannot be empty")
 	}
-	if len(r.Kind) == 0 {
+	if r.isKindEmpty() {
 		return fmt.Errorf("kind cannot be empty")
 	}
-
-	if len(r.Resource) == 0 {
-		r.Resource = flect.Pluralize(strings.ToLower(r.Kind))
-	}
-
+	// Check if the Group has a valid value for for it
 	if err := IsDNS1123Subdomain(r.Group); err != nil {
 		return fmt.Errorf("group name is invalid: (%v)", err)
 	}
-
-	r.GroupImportSafe = strings.Replace(r.Group, "-", "", -1)
-
+	// Check if the version is a valid value
 	versionMatch := regexp.MustCompile("^v\\d+(alpha\\d+|beta\\d+)?$")
 	if !versionMatch.MatchString(r.Version) {
 		return fmt.Errorf(
 			"version must match ^v\\d+(alpha\\d+|beta\\d+)?$ (was %s)", r.Version)
 	}
+	// Check if the Kind is a valid value
 	if r.Kind != flect.Pascalize(r.Kind) {
 		return fmt.Errorf("kind must be PascalCase (expected %s was %s)", flect.Pascalize(r.Kind), r.Kind)
 	}
 
+	// todo: move it for the proper place since they are not validations and then, should not be here
+	// Add in r.Resource the Kind plural
+	if len(r.Resource) == 0 {
+		r.Resource = flect.Pluralize(strings.ToLower(r.Kind))
+	}
+	// Replace the caracter "-" for "" to allow scaffold the go imports
+	r.GroupImportSafe = strings.Replace(r.Group, "-", "", -1)
 	return nil
```

Comment:
@mengqiy just note that I move the impl to a better order as ; check if is empty, then check if are valid values.
Also, add a TODO for we refactory it after in another PR since here should have just validations and not sets for the auxiliar values.

Reply:
I completely agree with you in this point, the `Validate` method is doing too many things. I was thinking on using constructor-oriented design (use a function called `New` or `NewXxxx` to create the instances of each type instead of literal struct constructs) and this not only applies here, but in a few more places.